core/src/container.cpp

```diff
@@ -790,60 +805,144 @@ bool Alternatives::canCompute() const {
-		StagePrivate* child = active_child_->pimpl();
-		if (child->canCompute())
-			return true;
-
-		// active child failed, continue with next
-		auto next = child->it();
-		++next;
-		if (next == pimpl()->children().end())
-			active_child_ = nullptr;
-		else
-			active_child_ = next->get();
+	auto impl { pimpl() };
+
+	if (impl->requiredInterface() == GENERATE) {
+		// current_generator_ is fixed if it produced solutions before
+		if (!solutions().empty())
+			return (*impl->current_generator_)->pimpl()->canCompute();
+		else {
+			// move to first generator that can run
+			while(impl->current_generator_ != impl->children().end() && !(*impl->current_generator_)->pimpl()->canCompute()) {
+				ROS_DEBUG_STREAM_NAMED("Fallbacks", "Generator '" << (*impl->current_generator_)->name() << "' can't compute, trying next one.");
+				++impl->current_generator_;
+			}
+			return impl->current_generator_ != impl->children().end();
+		}
 	}
-	return false;
+	else
+		return !impl->pending_states_.empty();
 }
 
 void Fallbacks::compute() {
-	if (!active_child_)
-		return;
+	auto impl { pimpl() };
 
-	try {
-		active_child_->pimpl()->runCompute();
-	} catch (const Property::error& e) {
-		active_child_->reportPropertyError(e);
-	}
+	if(impl->requiredInterface() == GENERATE)
+		impl->computeGenerate();
+	else
+		impl->computeFromExternal();
 }
 
 void Fallbacks::onNewSolution(const SolutionBase& s) {
 	liftSolution(s);
 }
 
+FallbacksPrivate::FallbacksPrivate(Fallbacks* me, const std::string& name)
+   : ParallelContainerBasePrivate(me, name) {
+	current_stage_ = nullptr;
+}
+
+void FallbacksPrivate::initializeExternalInterfaces() {
+	if (requiredInterface() & READS_START)
+		starts().reset(new Interface([this](Interface::iterator external, bool updated) {
+		                  this->onNewExternalState<Interface::FORWARD>(external, updated);
+		               }));
+	if (requiredInterface() & READS_END)
+		ends().reset(new Interface([this](Interface::iterator external, bool updated) {
+		                this->onNewExternalState<Interface::BACKWARD>(external, updated);
+		             }));
+}
+
+void FallbacksPrivate::onNewFailure(const Stage& /*child*/, const InterfaceState* /*from*/, const InterfaceState* /*to*/) {
+	// This override is deliberately empty.
+	// The method prunes solution paths when a child failed to find a valid solution for it,
+	// but in Fallbacks the next child might still yield a successful solution
+	// Thus pruning must only occur once the last child is exhausted (inside computeFromExternal)
+}
+
+void FallbacksPrivate::computeGenerate() {
+	assert(current_generator_ != children().end());
+	(*current_generator_)->pimpl()->runCompute();
+}
+
+template <typename Interface::Direction dir>
+void FallbacksPrivate::onNewExternalState(Interface::iterator external, bool updated) {
+	// TODO(v4hn): updated is not implemented
+	if(updated){
+		ROS_DEBUG_NAMED("Fallbacks", "updating external states is not supported in Fallbacks");
+		return;
+	}
```

Comment:
I'm wondering, why you didn't use [`ordered<ExternalState>`](https://github.com/ros-planning/moveit_task_constructor/blob/0aff5d56dd45b91b7c6c7c63c9674eec84e49aed/core/include/moveit/task_constructor/cost_queue.h#L28-L129). This way, incoming states would be ordered automatically and of course you could forward updated costs from the external state to the internal ones.

Reply:
I just pushed a change to `ordered<...>`, as proposed. I didn't manage to implement the corresponding cost updates yet though. Additionally I just noticed Fallbacks should not push pruned states to children either.

My intuition for the cyclic queue was that I would expect early stages in a `Fallbacks` container to be cheaper to compute, so computing the cheaper stages for all states instead of continuing the best-first-search through all children might be more economical. Either way should work reasonably well though.